pkg/proxy/userspace/roundrobin.go

```diff
@@ -245,16 +245,19 @@ func (lb *LoadBalancerRR) updateAffinityMap(svcPort proxy.ServicePortName, newEn
 
 // buildPortsToEndpointsMap builds a map of portname -> all ip:ports for that
 // portname. Expode Endpoints.Subsets[*] into this structure.
-func buildPortsToEndpointsMap(endpoints *v1.Endpoints) map[string][]hostPortPair {
-	portsToEndpoints := map[string][]hostPortPair{}
+func buildPortsToEndpointsMap(endpoints *v1.Endpoints) map[string][]string {
+	portsToEndpoints := map[string][]string{}
 	for i := range endpoints.Subsets {
 		ss := &endpoints.Subsets[i]
 		for i := range ss.Ports {
 			port := &ss.Ports[i]
 			for i := range ss.Addresses {
 				addr := &ss.Addresses[i]
-				portsToEndpoints[port.Name] = append(portsToEndpoints[port.Name], hostPortPair{addr.IP, int(port.Port)})
-				// Ignore the protocol field - we'll get that from the Service objects.
+				hpp := &hostPortPair{addr.IP, int(port.Port)}
+				if isValidEndpoint(hpp) {
```

Comment:
You should just make `isValidEndpoint` take an IP and port. And then I believe the entire hostPortPair type would be unused. Also, `flattenValidEndpoints` would be unused (other than the test case in `roundrobin_test.go` which is pointless now since nothing in the real code is using that function). So it could also be removed. And as Clayton said before, there ought to be a unit test of `buildPortsToEndpointsMap`.

Also, `pkg/proxy/winuserspace/` uses exactly the same code and should get exactly the same fix